File: sdk/clojure/src/bsdkrun/sandbox.clj
(ns bsdkrun.sandbox
  "The machine lifecycle. A \"sandbox\" here is just a plain map, e.g.
  `{:id \"abc123\" :ssh-port 2222}` — there is no object, no class. Create
  one with [[create!]], reconnect with [[get]], or enumerate with [[list]].

  Every function below that acts on a machine takes a `ref` first: a sandbox
  map, or a bare machine id/name string (see [[id]]) — `bsdkrun` itself
  resolves a bare id prefix or exact name (`core/src/db.rs`'s
  `find_machine`), so `(sandbox/stop! \"web-1\")` needs no lookup first. And
  since every `ref`-taking function returns either its result or (for
  lifecycle ops with nothing interesting to return) `ref` itself, they thread
  with `->`/`doto`:

  ```clojure
  (-> (sandbox/get \"web-1\")
      sandbox/start!
      (sandbox/exec! [\"uname\" \"-a\"])
      :stdout)

  (doto (sandbox/get \"web-1\")   ; same vm through every step, vm back at the end
    sandbox/start!
    (sandbox/exec! [\"setup.sh\"])
    sandbox/stop!)
  ```

  Mirrors `sdk/ruby/lib/bsdkrun/sandbox.rb`."
  (:refer-clojure :exclude [get list])
  (:require [clojure.string :as str]
            [bsdkrun.args :as args]
            [bsdkrun.errors :as errors]
            [bsdkrun.process :as process]
            [bsdkrun.types :as types]
            [bsdkrun.util :as util]))

(def ^:private id-re #"^[0-9a-f]{6,}$")
(def ^:private ssh-port-re #"ssh -p (\d+)")

(defn id
  "The machine id/name to hand the CLI for `ref` — a sandbox map's `:id`, or
  a bare id/name string, unchanged. Every function below that acts on a
  machine accepts either."
  [ref]
  (if (map? ref) (:id ref) ref))

(defn with-volume
  "Set the persistent volume a machine's rootfs is cloned onto/reused from
  (`-v`/`--volume`) on a *create-options* map — `:volume` is a create-time
  choice, fixed for the machine's lifetime, so this composes with `->`
  before [[create!]], not after it:

  ```clojure
  (-> {:os :linux :image \"alpine\"}
      (sandbox/with-volume \"web\")
      sandbox/create!)
  ```"
  [opts volume]
  (assoc opts :volume volume))

(defn with-network
  "Join a *create-options* map's guest to a global network (`--network`) —
  merges into `:net` rather than replacing it, so it composes with other
  `:net` keys (e.g. `:ports`, `:mac`) already set on `opts`. Composes with
  `->` before [[create!]]:

  ```clojure
  (-> {:os :linux :image \"alpine\"}
      (sandbox/with-network \"devnet\")
      sandbox/create!)
  ```

  To move an *existing* machine between networks, use [[connect-network!]]
  instead (applies on its next [[start!]])."
  [opts network]
  (assoc-in opts [:net :network] network))

(defn create!
  "Boot a new microVM and return `{:id ... :ssh-port ...}` (`:ssh-port` is
  nil unless the boot banner reported one, which only BSD guests do).

  `opts` is a create-options map discriminated on `:os` — see
  `bsdkrun.args/build-create-args`.

  Throws `errors/command-failed` if boot fails or no machine id is printed."
  [opts]
  (let [argv (args/build-create-args opts)
        res (process/run argv {:log-level (:log-level opts 1)})]
    (when-not (zero? (:exit-code res))
      (throw (errors/command-failed (assoc res :command "bsdkrun create"))))
    ;; Detached runs print just the machine id on stdout.
    (let [id (->> (str/split (:stdout res) #"\n")
                  (map str/trim)
                  (filter #(re-matches id-re %))
                  last)]
      (when-not id
        (throw (errors/command-failed
                (assoc res :command "bsdkrun create (no machine id in output)"))))
      (let [m (re-find ssh-port-re (:stderr res))]
        {:id id :ssh-port (when m (Long/parseLong (second m)))}))))

(defn list
  "List machines. `{:all true}` includes exited ones (default running only).

  Returns a vector of sandbox-info maps (see
  `bsdkrun.types/sandbox-info-from-row`)."
  ([] (list {}))
  ([opts]
   (let [argv (cond-> ["ps" "--json"] (:all opts) (conj "--all"))
         res (process/run! argv {:label "bsdkrun ps"})]
     (mapv types/sandbox-info-from-row (types/read-json-rows (:stdout res))))))

(defn- match-ref
  "Find the row in `rows` (sandbox-info maps) matching `ref-str` — exact name
  first (unambiguous, like `docker <name>`), then exact id, then a unique id
  prefix (Docker-style short ids). Mirrors `core/src/db.rs`'s
  `find_machine`."
  [rows ref-str]
  (or (some #(when (= (:name %) ref-str) %) rows)
      (some #(when (= (:id %) ref-str) %) rows)
      (some #(when (str/starts-with? (:id %) ref-str) %) rows)))

(defn get
  "Reconnect to an existing machine by id (a unique prefix is enough) or by
  exact name. `ref` is a sandbox map or a bare id/name string (see [[id]]).

  Throws `errors/sandbox-not-found` if nothing matches."
  [ref]
  (let [ref-str (id ref)
        found (match-ref (list {:all true}) ref-str)]
    (if found
      {:id (:id found)}
      (throw (errors/sandbox-not-found ref-str)))))

(defn exec!
  "Run a command in the guest through its exec agent.

  `command` may be a vector (argv, no shell parsing) or a bare string
  program name; with a string, `:args` supplies its arguments.

  `opts`:
    `:args`            arguments when `command` is a bare string
    `:env`             environment variables map (`-e K=V`)
    `:tty`             allocate a pseudo-TTY in the guest (`-t`)
    `:stdin`           data piped to the command's stdin
    `:cwd`             working directory, emulated via `sh -c 'cd ...'`
    `:throw-on-error`  throw `errors/command-failed` on a non-zero exit
    `:log-level`       per-command bsdkrun log level
    `:on-stdout`       called with each stdout byte-array chunk as it arrives
    `:on-stderr`       called with each stderr byte-array chunk as it arrives

  Returns `{:stdout ... :stderr ... :exit-code ... :command \"...\"}`."
  ([ref command] (exec! ref command {}))
  ([ref command {:keys [args env tty stdin cwd throw-on-error log-level on-stdout on-stderr]
                 :or {args [] env {} tty false throw-on-error false log-level 0}}]
   (let [argv0 (if (sequential? command) (vec command) (into [command] args))
         argv (if cwd
                (into ["/bin/sh" "-c" "cd \"$1\" && shift && exec \"$@\"" "sh" cwd] argv0)
                argv0)
         cli (cond-> ["exec"] tty (conj "-t"))
         cli (into cli (mapcat (fn [[k v]] ["-e" (str (name k) "=" v)]) env))
         cli (into cli (into [(id ref)] argv))
         res (process/run cli {:stdin stdin :log-level log-level
                               :on-stdout on-stdout :on-stderr on-stderr})
         result {:stdout (:stdout res)
                 :stderr (:stderr res)
                 :exit-code (:exit-code res)
                 :command (str "exec " (str/join " " argv))}]
     (if throw-on-error
       (types/throw-if-failed! result)
       result))))

(defn run-command!
  "Vercel-Sandbox-style alias for [[exec!]]: a program plus its args."
  ([ref command] (run-command! ref command [] {}))
  ([ref command args] (run-command! ref command args {}))
  ([ref command args opts]
   (exec! ref command (assoc opts :args args))))

(defn logs
  "Read the machine's console log. `{:boot true}` shows bsdkrun's own boot
  log instead of the console."
  ([ref] (logs ref {}))
  ([ref {:keys [boot]}]
   (let [argv (cond-> ["logs"] boot (conj "--boot"))
         argv (conj argv (id ref))]
     (:stdout (process/run argv)))))

(defn shell!
  "Attach an interactive shell to the machine (inherits the terminal).
  Returns true if the shell exited zero."
  [ref]
  (process/spawn-interactive! ["shell" (id ref)]))

(defn status
  "This machine's current status row, or nil if it's gone. `ref` may be a
  sandbox map or a bare id/name string."
  [ref]
  (match-ref (list {:all true}) (id ref)))

(defn running?
  "Whether the machine is currently running."
  [ref]
  (boolean (:running (status ref))))

(defn- lifecycle!
  "Run a fire-and-forget lifecycle CLI command, throwing on failure. Returns
  `ref` unchanged (not its result) so lifecycle calls compose with
  `->`/`doto` — see the namespace docstring."
  [ref argv label]
  (process/run! argv {:label label})
  ref)

(defn stop!
  "Stop the machine. BSD guests are cleanly powered off; Linux is SIGTERM'd."
  [ref]
  (lifecycle! ref ["stop" (id ref)] "bsdkrun stop"))

(defn start!
  "Restart a stopped machine in place (same id, disk/rootfs). Boots detached."
  [ref]
  (lifecycle! ref ["start" (id ref)] "bsdkrun start"))

(defn remove!
  "Remove the machine and its state. `{:force true}` stops it first if
  running."
  ([ref] (remove! ref {}))
  ([ref {:keys [force]}]
   (let [argv (cond-> ["rm"] force (conj "--force"))
         argv (conj argv (id ref))]
     (lifecycle! ref argv "bsdkrun rm"))))

(defn update!
  "Change the recorded vCPU / RAM. Applies on the next [[start!]]."
  ([ref] (update! ref {}))
  ([ref {:keys [cpus mem]}]
   (let [argv (cond-> ["update" (id ref)]
                (some? cpus) (into ["--cpus" (str cpus)])
                (some? mem) (into ["--mem" (str mem)]))]
     (lifecycle! ref argv "bsdkrun update"))))

(defn connect-network!
  "Join or switch this machine to a global network. Applies on next
  [[start!]]."
  [ref network]
  (lifecycle! ref ["network" "connect" (id ref) network] "bsdkrun network connect"))

(defn disconnect-network!
  "Detach this machine from its network. Applies on next [[start!]]."
  [ref]
  (lifecycle! ref ["network" "disconnect" (id ref)] "bsdkrun network disconnect"))

(defn- agent!
  "Run an in-guest agent CLI family (`ssh`, `tailscale`), throwing on
  failure."
  [ref family action {:keys [env] :or {env {}}}]
  (let [res (process/run (into [family (id ref)] action) {:env env})
        result {:stdout (:stdout res)
                :stderr (:stderr res)
                :exit-code (:exit-code res)
                :command (str family " " (str/join " " action))}]
    (types/throw-if-failed! result)))

(defn ssh-setup!
  "Install SSH keys in the guest via the agent (`ssh setup`). With no keys,
  the CLI installs your local `~/.ssh/*.pub`.

  `opts`: `:user` (target user, default root), `:key` (a literal key or
  `.pub` path, or a vector of them)."
  ([ref] (ssh-setup! ref {}))
  ([ref {:keys [user key]}]
   (let [action (cond-> ["setup"] user (into ["--user" user]))
         action (into action (mapcat (fn [k] ["--key" k]) (util/as-seq key)))]
     (agent! ref "ssh" action {}))))

(defn tailscale-up!
  "Put the guest on your tailnet (`tailscale setup`).

  `opts`: `:authkey` (tailnet auth key, sent as `TS_AUTHKEY`), `:hostname`
  (machine name on the tailnet), `:args` (extra args passed through to
  `tailscale up`)."
  ([ref] (tailscale-up! ref {}))
  ([ref {:keys [authkey hostname args]}]
   (let [action (cond-> ["setup"] hostname (into ["--hostname" hostname]))
         action (into action (or args []))]
     (agent! ref "tailscale" action {:env (if authkey {"TS_AUTHKEY" authkey} {})}))))

File: sdk/ruby/lib/bsdkrun/sandbox.rb
# frozen_string_literal: true

require "json"

module Bsdkrun
  # A handle to a running (or stopped) bsdkrun microVM.
  #
  # Create one with {Sandbox.create}, reconnect with {Sandbox.get}, or enumerate
  # with {Sandbox.list}.
  #
  # @example
  #   box = Bsdkrun::Sandbox.create(os: "linux", image: "alpine")
  #   box.exec(["uname", "-a"]).text
  #   box.stop
  class Sandbox
    ID_RE = /\A[0-9a-f]{6,}\z/
    SSH_PORT_RE = /ssh -p (\d+)/
    private_constant :ID_RE, :SSH_PORT_RE

    # The machine's Docker-style short id.
    # @return [String]
    attr_reader :id

    # Host port forwarded to the guest's SSH, if the boot banner reported one.
    # @return [Integer, nil]
    attr_reader :ssh_port

    # @param id [String]
    # @param ssh_port [Integer, nil]
    def initialize(id, ssh_port: nil)
      @id = id
      @ssh_port = ssh_port
    end

    # Read and write files in the guest.
    # @return [FileSystem]
    def fs
      @fs ||= FileSystem.new(@id)
    end

    class << self
      # Boot a new microVM and return a handle to it.
      #
      # Accepts create options as a keyword list or a Hash; discriminated on
      # +:os+ ("linux", "freebsd", "netbsd", "firmware", "kernel").
      #
      # @param opts [Hash]
      # @return [Sandbox]
      # @raise [CommandFailed] if boot fails or no machine id is printed.
      def create(opts = {}, **kwargs)
        opts = normalize(opts.merge(kwargs))
        args = Args.build_create_args(opts)
        res = Process.run(args, log_level: opts.fetch(:log_level, 1))
        if res.exit_code != 0
          raise CommandFailed.new(
            exit_code: res.exit_code, stdout: res.stdout, stderr: res.stderr,
            command: "bsdkrun create"
          )
        end
        # Detached runs print just the machine id on stdout.
        id = res.stdout.split("\n").map(&:strip).select { |l| l.match?(ID_RE) }.last
        unless id
          raise CommandFailed.new(
            exit_code: res.exit_code, stdout: res.stdout, stderr: res.stderr,
            command: "bsdkrun create (no machine id in output)"
          )
        end
        match = res.stderr.match(SSH_PORT_RE)
        new(id, ssh_port: match && match[1].to_i)
      end

      # Reconnect to an existing machine by id (a unique prefix is enough).
      #
      # @param id [String]
      # @return [Sandbox]
      # @raise [SandboxNotFound]
      def get(id)
        row = list(all: true).find { |m| m.id == id || m.id.start_with?(id) || m.name == id }
        raise SandboxNotFound, id unless row

        new(row.id)
      end

      # List machines. +all: true+ includes exited ones (default running only).
      #
      # @param all [Boolean]
      # @return [Array<SandboxInfo>]
      def list(all: false)
        args = ["ps", "--json"]
        args << "--all" if all
        res = Process.run!(args, label: "bsdkrun ps")
        rows = JSON.parse(res.stdout.empty? ? "[]" : res.stdout)
        rows.map { |row| SandboxInfo.from_row(row) }
      end

      # Symbolize keys and normalize a nested +:net+ hash.
      # @!visibility private
      def normalize(opts)
        h = opts.each_with_object({}) { |(k, v), acc| acc[k.to_sym] = v }
        h[:net] = h[:net].each_with_object({}) { |(k, v), acc| acc[k.to_sym] = v } if h[:net].is_a?(Hash)
        h
      end
    end

    # Run a command in the guest through its exec agent.
    #
    # +command+ may be an Array (argv, no shell parsing) or a String program
    # name; with a String, +args:+ supplies its arguments.
    #
    # @param command [String, Array<String>]
    # @param args [Array<String>] arguments when +command+ is a bare String.
    # @param env [Hash] environment variables (+-e K=V+).
    # @param tty [Boolean] allocate a pseudo-TTY in the guest (+-t+).
    # @param stdin [String, nil] data piped to the command's stdin.
    # @param cwd [String, nil] working directory (emulated via +sh -c 'cd …'+).
    # @param throw_on_error [Boolean] raise {CommandFailed} on a non-zero exit.
    # @param log_level [Integer] per-command bsdkrun log level.
    # @param on_stdout [Proc, nil] called with each stdout chunk as it arrives.
    # @param on_stderr [Proc, nil] called with each stderr chunk as it arrives.
    # @return [Result]
    def exec(command, args: [], env: {}, tty: false, stdin: nil, cwd: nil,
             throw_on_error: false, log_level: 0, on_stdout: nil, on_stderr: nil)
      argv = command.is_a?(Array) ? command.dup : [command, *args]

      if cwd
        # Emulate a working directory: cd, drop it, then exec the real argv.
        argv = ["/bin/sh", "-c", 'cd "$1" && shift && exec "$@"', "sh", cwd, *argv]
      end

      cli = ["exec"]
      cli << "-t" if tty
      env.each { |k, v| cli.push("-e", "#{k}=#{v}") }
      cli.push(@id, *argv)

      res = Process.run(cli, stdin: stdin, log_level: log_level,
                        on_stdout: on_stdout, on_stderr: on_stderr)
      result = Result.new(
        stdout: res.stdout, stderr: res.stderr, exit_code: res.exit_code,
        command: "exec #{argv.join(' ')}"
      )
      result.throw_if_failed! if throw_on_error
      result
    end

    # Vercel-Sandbox-style alias for {#exec}: a program plus its args.
    #
    # @param command [String]
    # @param args [Array<String>]
    # @return [Result]
    def run_command(command, args = [], **opts)
      exec(command, args: args, **opts)
    end

    # Read the machine's console log.
    #
    # @param boot [Boolean] show bsdkrun's own boot log instead of the console.
    # @return [String]
    def logs(boot: false)
      args = ["logs"]
      args << "--boot" if boot
      args << @id
      Process.run(args).stdout
    end

    # Attach an interactive shell to the machine (inherits the terminal).
    # @return [Boolean] true if the shell exited zero.
    def shell
      Process.spawn_interactive(["shell", @id])
    end

    # Fetch this machine's current status row, or nil if it's gone.
    # @return [SandboxInfo, nil]
    def status
      Sandbox.list(all: true).find { |m| m.id == @id }
    end

    # Whether the machine is currently running.
    # @return [Boolean]
    def running?
      s = status
      s ? s.running : false
    end

    # Stop the machine. BSD guests are cleanly powered off; Linux is SIGTERM'd.
    # @return [void]
    def stop
      lifecycle(["stop", @id], "bsdkrun stop")
    end

    # Restart a stopped machine in place (same id, disk/rootfs). Boots detached.
    # @return [void]
    def start
      lifecycle(["start", @id], "bsdkrun start")
    end

    # Remove the machine and its state. +force+ stops it first if running.
    # @param force [Boolean]
    # @return [void]
    def remove(force: false)
      args = ["rm"]
      args << "--force" if force
      args << @id
      lifecycle(args, "bsdkrun rm")
    end

    # Change the recorded vCPU / RAM. Applies on the next {#start}.
    # @param cpus [Integer, nil]
    # @param mem [Integer, nil]
    # @return [void]
    def update(cpus: nil, mem: nil)
      args = ["update", @id]
      args.push("--cpus", cpus.to_s) unless cpus.nil?
      args.push("--mem", mem.to_s) unless mem.nil?
      lifecycle(args, "bsdkrun update")
    end

    # Join or switch this machine to a global network. Applies on next {#start}.
    # @param network [String]
    # @return [void]
    def connect_network(network)
      lifecycle(["network", "connect", @id, network], "bsdkrun network connect")
    end

    # Detach this machine from its network. Applies on next {#start}.
    # @return [void]
    def disconnect_network
      lifecycle(["network", "disconnect", @id], "bsdkrun network disconnect")
    end

    # Install SSH keys in the guest via the agent (+ssh setup+). With no keys,
    # the CLI installs your local +~/.ssh/*.pub+.
    #
    # @param user [String, nil] target user (default root).
    # @param key [String, Array<String>, nil] literal key(s) or +.pub+ path(s).
    # @return [Result]
    def ssh_setup(user: nil, key: nil)
      action = ["setup"]
      action.push("--user", user) if user
      Array(key).each { |k| action.push("--key", k) }
      agent("ssh", action)
    end

    # Put the guest on your tailnet (+tailscale setup+).
    #
    # @param authkey [String, nil] tailnet auth key (sent as +TS_AUTHKEY+).
    # @param hostname [String, nil] machine name on the tailnet.
    # @param args [Array<String>] extra args passed through to +tailscale up+.
    # @return [Result]
    def tailscale_up(authkey: nil, hostname: nil, args: [])
      action = ["setup"]
      action.push("--hostname", hostname) if hostname
      action.concat(args)
      agent("tailscale", action, env: authkey ? { "TS_AUTHKEY" => authkey } : {})
    end

    private

    # Run a fire-and-forget lifecycle CLI command, raising on failure.
    def lifecycle(args, label)
      Process.run!(args, label: label)
      nil
    end

    # Run an in-guest agent CLI family (+ssh+, +tailscale+), raising on failure.
    def agent(family, action, env: {})
      res = Process.run([family, @id, *action], env: env)
      result = Result.new(
        stdout: res.stdout, stderr: res.stderr, exit_code: res.exit_code,
        command: "#{family} #{action.join(' ')}"
      )
      result.throw_if_failed!
    end
  end
end
